py-rse/style.Rmd

```diff
@@ -4,207 +4,255 @@
     Section \@ref(py-rse-package-py-import).
 
-### Do not comment and uncomment sections of code to change behavior.
+That much is clear,
+but programmers disagree (strongly) on whether high-level functions should come first or last,
+i.e.,
+whether `main` should be the first function in the file or the last one.
+
+We prefer the former approach,
+i.e.,
+putting high-level functions at the top of the file.
+The reason is that when people encounter code for the first time,
+they usually scan it from top to bottom.
+The first function they read should therefore help them understand
+what the rest of the file contains and where to look for it.
+Three common patterns that people use for this are:
+
+1.  Figure out what the user has asked it to do (Chapter \@ref(py-rse-configuration)).
+2.  Read all input data.
+3.  Process it.
+4.  Write output.
 
-If you need to do something in some runs of the program and not do it in others,
-use an `if` statement to enable or disable that block of code.
-It's much more reliable—you're far less likely to accidentally comment out one too many lines—and
-you may find that you want to leave those conditional sections in the finished program
-for logging purposes (Chapter \@ref(logging)).
+or:
 
-### Keep functions short.
+1.  Figure out what the user has asked for.
+2.  For each input file:
+    1.  Read.
+    2.  Process.
+    3.  Write file-specific output (if any).
+3.  Write summary output (if any).
 
-Nothing should be more than a page long
-or have more than three levels of indentation because of nested loops and conditionals.
-Anything longer or more deeply nested will be hard for readers to fit into [working memory][working-memory];
-when you find yourself needing to break these limits,
-extract a function.
+or:
 
-### Handle special cases at the start of the function.
+1.  Figure out what the user has asked for.
+2.  Repeatedly:
+    1.  Wait for user input.
+    2.  Do what the user has asked.
+3.  Exit when a "stop" command of some sort is received.
 
-This helps readers mentally get them out of the way
-and focus on the "normal" case.
+Each step in each of the outlines above usually becomes a function.
+Those functions depend on others,
+some of which are written to break code into comprehensible chunks and are then called just once,
+others of which are utilities that may be called many times from many different places.
+We put all of the single-use functions in the first half of the file
+in the order in which they are likely to be called,
+and then put all of the multi-use utility functions in the bottom of the file in alphabetical order.
+If any of those utility functions are used by other scripts or programs,
+they should go in a file of their own.
+In fact,
+this is a good practice even if they're only used by one program,
+since it signals even more clearly which functions are specific to this program
+and which are likely to be reused elsewhere.
```

Comment:
Should we include anything about breaking into multiple files?

Reply:
We might also want to discuss how the code we've developed in the preceding chapters (i.e. 'countwords.py', 'collate.py' and 'plotcounts.py', plus the soon to be renamed `mymodule.py`) applies these patterns.

Reply:
How to break code between files feels like it belongs somewhere near here, but I'm worried that this chapter is already a long catalog that readers won't be able to absorb. Could it go with packaging?

Reply:
I was thinking something short, a sentence or two, but I am also fine with leaving as is since we do mention when to break into files. If we were to put it elsewhere, I think it would go best with where we create the second script together with an explanation of why we didn't just continue on the first script.

Reply:
Can you point me at where we mention breaking code into files? And which script do you mean by "second script"?

Reply:
Above at line 250-251:

> If any of those utility functions are used by other scripts or programs,
> they should go in a file of their own.

By second script I was referring to when we in the Python scripting chapter decide to create a new file for some of the functionality instead of keep adding to the first file we created. In the end we have four files (countwords, collate, plotcounts, mymodule), but we could have had one really long file.

Reply:
Gotcha - added